### PR TITLE
Give informative message for parsing errors

### PR DESCRIPTION
**Pull request by timholy**
Fixes https://github.com/JuliaCI/Coverage.jl/issues/299
Closes #42

### CODE REVIEW

File: src/CoverageTools.jl [hunk 1]
```diff
@@ -174,32 +174,38 @@ module CoverageTools
...
-            if !isempty(flines)
-                flines .+= lineoffset
-                for l in flines
-                    if l > length(coverage)
-                        resize!(coverage, l)
-                    end
-                    if coverage[l] === nothing
-                        coverage[l] = 0
+                # now we can parse the next chunk of the input
+                ast, pos = Meta.parse(content, pos)
```
Comment (vtjnash) [inline src/CoverageTools.jl]:
```suggestion
                ast, pos = Meta.parse(content, pos, raise=false)
```

File: src/CoverageTools.jl [hunk 2]
```diff
@@ -174,32 +174,38 @@ module CoverageTools
...
+                            resize!(coverage, l)
+                        end
+                        if coverage[l] === nothing
+                            coverage[l] = 0
+                        end
                     end
                 end
             end
+        catch err
+            errkind = isa(err, Meta.ParseError) ? "parsing error" : "error"
```
Comment (fingolfin) [inline src/CoverageTools.jl]: Alternatively, we could call `Meta.parse` with `raise=false` and then after `isa(ast, Expr) || continue` check for `ex.head === :error` to detect an error. Advantage: we then could use `pos` to determine the location of the parsing error more precisely (`pos` will point just *past* the problematic code, but I guess one could say something like "parsing error was before line NN, position PP".

File: src/CoverageTools.jl [hunk 3]
```diff
@@ -186,8 +186,12 @@ module CoverageTools
             lineoffset = searchsortedlast(linepos, pos - 1) - 1
 
             # now we can parse the next chunk of the input
-            ast, pos = Meta.parse(content, pos)
+            ast, pos = Meta.parse(content, pos; raise=false)
             isa(ast, Expr) || continue
+            if ast.head ∈ (:error, :incomplete)
+                lineoffset = searchsortedlast(linepos, pos - 1) - 1
+                throw(Base.Meta.ParseError("parsing error in $(fc.filename), successfully handled up to line $lineoffset"))
```
Comment (fingolfin) [inline src/CoverageTools.jl]: I would avoid using the identifier `lineoffset` here as that's used before to store a 0-based offset while we want to print a 1-based line number to the user).

Also, I think we could do even better by extracting the error message.
```
julia> str = """for x =

        end"""; Meta.parse(str, 1, raise = false)
(:($(Expr(:error, "unexpected \"end\""))), 13)


julia> str = """for x =

        """; Meta.parse(str, 1, raise = false)
(:($(Expr(:incomplete, "incomplete: premature end of input"))), 10)
```

So how about this:
```suggestion
                line = searchsortedlast(linepos, pos - 1)
                throw(Base.Meta.ParseError("parsing error in $(fc.filename):$line: $(ast.args[1])"))
```
Or if we want to match the message Julia prints more closely:
```suggestion
                line = searchsortedlast(linepos, pos - 1)
                throw(Base.Meta.ParseError("syntax: $(ast.args[1]) at $(fc.filename):$line"))
```

Well, except the error message actually can contain a "filename" already... ugh:
```
julia> str = """
       function parseerr()
           s = 0
           for i = [1,2,3]   # this line has a parsing error
               s += i
           end
           return s
       """;

julia> Meta.parse(str, 1, raise = false)
(:($(Expr(:incomplete, "incomplete: \"function\" at none:1 requires end"))), 121)
```
Interestingly, in the Julia master branch the lower-level parsing functions all have kwargs `filename` and `index` -- but `parse` doesn't expose them. Guess it's time for yet another PR :-).

**Review by fingolfin — APPROVED** [review]

File: test/runtests.jl [hunk 4]
```diff
@@ -188,6 +188,16 @@ end
...
     end
+
+    srcdir = joinpath(@__DIR__, "BustedPackage", "src")
+    clean_folder(srcdir)
+    cmdstr = "using BustedPackage; BustedPackage.greet()"
+    cd("BustedPackage") do
+        run(`$(Base.julia_cmd()) --startup-file=no --code-coverage=user --project -e $cmdstr`)
+    end
+    bustedfile = joinpath(srcdir, "parseerr.jl")
+    @test_throws Base.Meta.ParseError("parsing error in $bustedfile, successfully handled up to line 6") process_file(bustedfile, srcdir)
```
Comment (fingolfin) [inline test/runtests.jl]:
```suggestion
    @test_throws Base.Meta.ParseError("parsing error in $bustedfile:7: space before \"[\" not allowed in \"i [\"") process_file(bustedfile, srcdir)
```

Comment (fingolfin) [reply test/runtests.jl]: @timholy sorry my change proposal was not complete, the test needs to be adapted, too (the above might work, but I couldn't test it yet).

Thank you for your patience and willingness to help. If you prefer me to fix the test (you already went way beyond anything to be expected), let me know, I'll do it later today.

File: test/runtests.jl [hunk 5]
```diff
@@ -188,6 +188,16 @@ end
...
     end
+
+    srcdir = joinpath(@__DIR__, "BustedPackage", "src")
+    clean_folder(srcdir)
+    cmdstr = "using BustedPackage; BustedPackage.greet()"
+    cd("BustedPackage") do
+        run(`$(Base.julia_cmd()) --startup-file=no --code-coverage=user --project -e $cmdstr`)
+    end
+    bustedfile = joinpath(srcdir, "parseerr.jl")
+    @test_throws Base.Meta.ParseError("parsing error in $bustedfile:7: space before \"[\" not allowed in \"i [\"") process_file(bustedfile, srcdir)
```
Comment (fingolfin) [inline test/runtests.jl]: Annoyingly the message differs between older Julia versions and 1.5 (or so), where the string " at none:4" is added to the message. 
```suggestion
    msg = "parsing error in $bustedfile:7: space before \"[\" not allowed in \"i [\""
    if VERSION >= v"1.5"
        msg *= " at none:4"
    end
    @test_throws Base.Meta.ParseError(msg) process_file(bustedfile, srcdir)
```

**Comment by timholy** [comment]
Thanks for fixing it!

**Comment by fingolfin** [comment]
Thanks for making the PR :-) Very nice improvement

**Pull request merged**